cmakelib/lexer/lexer.go

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package lexer implements a participle Lexer for the CMakeLists.txt language.
+// See https://cmake.org/cmake/help/v3.0/manual/cmake-language.7.html for more details.
+package lexer
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/alecthomas/participle/lexer"
+)
+
+const (
+	_ rune = lexer.EOF - iota
+	Space
+	Newline
+	EscapeSequence
+	Quote
+	Quoted
+	BracketOpen
+	BracketContent
+	BracketClose
+	VarOpen
+	VarClose
+	Unquoted
+	Punct
+	Comment
+)
+
+type tokenDefinition struct {
+	kind rune
+	name string
+	pat  string
+}
+
+// Ordered token definitions. Regular expressions are matched in order!
+// Empty patterns are omitted.
+var tokenDefs = []tokenDefinition{
+	{lexer.EOF, "EOF", ``},
+	{Space, "Space", `[ \t]+`},
+	{Newline, "Newline", `\n+`},
+	{EscapeSequence, "EscapeSequence", `(?s:\\.)`},
+	{Quote, "Quote", `"`},
+	{Quoted, "Quoted", ``},
+	{BracketOpen, "BracketOpen", `\[=*\[`},
+	{BracketContent, "BracketContent", ``},
+	{BracketClose, "BracketClose", `]=*]`},
+	{VarOpen, "VarOpen", `\$\w*\{`},
+	{VarClose, "VarClose", `}`},
+	{Punct, "Punct", `[()]`},
+	{Comment, "Comment", `#`},
+	// Unquoted is the last match, so can start with a broader set of characters,
+	// but cannot consume additional characters which would match above.
+	// We could also only every have it match a single character at a time.
+	{Unquoted, "Unquoted", `[^\s()#"][^][\$\s()#"\\}]*`},
+}
+
+var (
+	eolBytes   = []byte("\n")
+	tokenSyms  map[string]rune
```

Comment:
You could (if you like) consolidate these with their initialization:

```go
tokenSyms  = make(map[string]rune)
tokenNames = make(map[rune]string)
…
```

and save a couple lines in `init`. YMMV.